hubble-relay: Add node status message

gandro opened the pull request:
This adds a status message to the observer service implementation
of hubble-relay. The status message is added as a new variant of
the GetFlowsResponse, therefore older clients will silently ignore
it.

The new node status event is defined as follow:

```protobuf
// NodeStatusEvent is a message sent by hubble-relay to inform clients about
// the state of a particular node.
message NodeStatusEvent {
    // state_change contains the new node state
    NodeState state_change = 1;
    // node_names is the list of nodes for which the above state changes applies
    repeated string node_names = 2;
    // message is an optional message attached to the state change (e.g. an
    // error message). The message applies to all nodes in node_names.
    string message = 3;
}

enum NodeState {
    // UNKNOWN_NODE_STATE indicates that the state of this node is unknown.
    UNKNOWN_NODE_STATE = 0;
    // NODE_CONNECTED indicates that we have established a connection
    // to this node. The client can expect to observe flows from this node.
    NODE_CONNECTED = 1;
    // NODE_UNAVAILABLE indicates that the connection to this
    // node is currently unavailable. The client can expect to not see any
    // flows from this node until either the connection is re-established or
    // the node is gone.
    NODE_UNAVAILABLE = 2;
    // NODE_GONE indicates that a node has been removed from the
    // cluster. No reconnection attempts will be made.
    NODE_GONE = 3;
    // NODE_ERROR indicates that a node has reported an error while processing
    // the request. No reconnection attempts will be made.
    NODE_ERROR = 4;
}
```

The implementation makes sure to send down the initial node status
for all nodes participating in the request first, i.e. before any flows.

The node state `GONE` is not used in this initial version, as the
current implementation of the peer management in hubble-relay does not
inform the running GetFlows request about the removal of a node.
We mark any failed node as `UNAVAILABLE` for now.

**Review per commit** as this PR also performs a bit of cleanup.

Fixes: #11360

rolinh commented on api/v1/observer/observer.proto:
```diff
@@ -86,10 +86,26 @@ message GetFlowsRequest {
 
+// RelayStatus is a message sent by hubble-relay to indicate which nodes
+// were unresponsive during the request.
+message RelayStatus {
```
Thinking: this message definition seems generic enough that it could be reused in other gRPC APIs. I'm mostly thinking about the relay API proposal in #11026. With this in mind, what about moving the definition to `api/v1/relay/relay.proto`? And then maybe rename it to simply `Status` as `relay.RelayStatus` would clutter a bit.

gandro replied:
Good suggestion!

rolinh commented on pkg/hubble/relay/observer.go:
```diff
@@ -31,6 +33,21 @@ import (
+func relayStatusResponse(numPeers int, failedPeers []string) *observerpb.GetFlowsResponse {
+	return &observerpb.GetFlowsResponse{
+		Time:     ptypes.TimestampNow(),
+		NodeName: node.GetName(),
```
Does the node name makes sense in the context of hubble-relay? Well, I guess we have to fill this anyway and I see no better alternative.

rolinh commented on pkg/hubble/relay/observer.go:
```diff
@@ -54,12 +71,55 @@ func (s *Server) GetFlows(req *observerpb.GetFlowsRequest, stream observerpb.Obs
+	}()
+
+	go func() {
+		defer close(flows)
```
I guess that closing the flows channel here is assumed to be OK even though there are other routines that write into this channel because this goroutine stops only when the context is done, right? I'm not sure it's safe though because if this one returns on `ctx.Done()` before one of the other writers, I think there could be a chance for a tiny race that could lead to a panic (writing to closed channel).

gandro replied:
You are correct. I have not considered the case where the `ctx.Done()` is racy.

The basic "shutdown" procedure is intended to be as follows (in non-follow-mode):
 - Wait until each each peer in the errgroup finishes its request, this ensures we have collected all `failedPeers`. We indicate this by closing the `failedPeers` channel.
-  The status go-routine sees that it has received all failed peers, therefore it can generate the `RelayStatus` message with the complete list of peers. It then closes `flows`.

In follow-mode, we do not have to option to send back a final status message, since these kind of clients just disconnect. Therefore, we will eventually just get `io.EOF` from `stream.Send` which will cancel the context and tear down all go routines.

But yes, as it is currently we have the problem that we might write to closed channels. I will rework the logic to avoid this.

rolinh replied:
Here is an idea to try and simplify this although I didn't try to implement the logic myself:
What about getting rid of the `failedPeers` channel altogether and leverage your new `RelayStatus` message instead? So when a peer fails, instead of sending `p.Name` in the `failedPeers` chan, you just send a `GetFlowsResponse` with a `RelayStatus` into the flows channel with information about the peer that failed. These status results can then be aggregated by the receiver and sent when appropriate (eg when the timer ticks); it's just a matter of checking for them when reading from the `flows` channel. This way you don't have to create an additional goroutine and can leave the tear-down logic as it was before.

rolinh reviewed (commented):
Very nice improvement overall! I'm a little worried about the complexity of the `GetFlows` implementation. Eventually, it should be refactored so that it could be more readable/testable/debuggable (work for a follow-up PR though).

gandro committed:
hubble-relay: Split out GetFlows logic into functions

This commit contians no functional change.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api: Add node status message for hubble-relay

This introduces a new  node status message sent hubble-relay. The
message is used to inform clients about the connectivity of nodes
participating in a request.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble-relay: Send node status events in GetFlows

This adds support for NodeStatusEvents in the GetFlows RPC call.
The intent of this event is to inform downstream consumers about the
state of the nodes which are participating in the current request.

The node state `NODE_GONE` is not used in this initial version, as the
current implementation of the peer management in hubble-relay does not
inform the running GetFlows request about the removal of a node.

If a peer is not ready when the request is started, we mark it as
`NODE_UNAVAILABLE`. If a peer errors out during a request, we indicate
this via `NODE_ERROR` and propagate the received error message.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

rolinh commented on api/v1/relay/relay.proto:
```diff
@@ -0,0 +1,42 @@
+
+enum NodeState {
+    // UNKNOWN_NODE_STATE indicates that the state of this node is unknown.
+    UNKNOWN_NODE_STATE = 0;
```
Since the enum itself is called `NodeState`, I don't think it's necessary to repeat `NODE_STATE` in each state name.
They could just be `UNKNOWN`, `CONNECTED`, `UNAVAILABLE` and `GONE`.

gandro replied:
UNKNOWN needs to be prefixed, because protobuf does not have enum namespaces and will therefore collide with any other enum which will have the UNKNOWN variant. But I guess for the others, we can do that.

gandro replied:
I have shortened the variants to `NODE_xxx`. I think a prefix is still useful, since `relay.proto` will gain quite a few other enums that maybe also want to make use of variants like `ERROR`.

rolinh commented on pkg/hubble/relay/observer.go:
```diff
@@ -20,17 +20,111 @@ import (
+	}
+}
+
+func sortFlows(
```
As this is the only place where `sortedFlows` is written to, what about changing the prototype to:
```go
func sortFlows(
	ctx context.Context,
	flows <-chan *observerpb.GetFlowsResponse,
	qlen int,
	bufferDrainTimeout time.Duration,
) <-chan *observerpb.GetFlowsResponse {...}
```
And have `sortFlows` be responsible of creating and closing the channel (this requires spawning a goroutine in the func body rather than leaving this work to the caller)? I think the code would look a little cleaner from the caller's perspective as instead of this:
```go
       sortedFlows := make(chan *observerpb.GetFlowsResponse, qlen)
        go func() {
                defer close(sortedFlows)
                // sortFlows returns either when flows is closed or ctx expires
                sortFlows(ctx, flows, qlen, s.opts.BufferDrainTimeout, sortedFlows)
        }()
```
You'd have this:
```go
sortedFlows := sortFlows(ctx, flows, qlen, s.opts.BufferDrainTimeout)
```

rolinh reviewed (commented):
I think the changes you made to the API proposal are good. However, I need to experiment/play a bit more with the API in order to understand what could be missing or improved before giving a final approval.

gandro committed:
hubble-relay: Suppress duplicate errors within the same time window

This commit adds a new stage to the hubble-relay processing which
supresses duplicate errors which may occur on multiple nodes within a
certain time window.

The list of nodes is merged, such that the reported error contains
each node on which the error occured.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

rolinh commented on pkg/hubble/relay/relayoption/defaults.go:
```diff
@@ -24,11 +24,13 @@ import (
+
+	BufferMaxLen:           100,
+	BufferDrainTimeout:     1 * time.Second,
+	ErrorAggregationWindow: 3 * time.Second,
```
Was this value empirically chosen? It seems a little short to me.

gandro replied:
For small cluster sizes, 3 seconds seems to work nicely: It's long enough such that each node will have a chance to report their error, but short enough so the clients don't have too long for errors to be reported.

But I actually don't know what a good window size would be for larger clusters. Too short and we end up producing unnecessary traffic, too long and we risk merging uncorrelated errors (and clients have to wait long for errors as well).

gandro replied:
Bumped to 10sec

rolinh reviewed (commented):
I like the `NODE_ERROR` change, very welcome! As a follow-up to this PR, I wonder if we should not also record the last error for every peer along with the corresponding timestamp so that we could then serve this information via a corresponding note status grpc service.

rolinh commented on api/v1/relay/relay.proto:
```diff
@@ -0,0 +1,48 @@
+    // state_change contains the new node state
+    NodeState state_change = 1;
+    // node_name is the list of nodes for which the above state changes applies
+    repeated string node_name = 2;
```
I think it should be pluralized. This already confused me while reviewing https://github.com/cilium/hubble/pull/260 as I thought that `NodeName` was just a node name when in fact it's a list of node names.

rolinh commented on pkg/hubble/relay/observer.go:
```diff
@@ -20,17 +20,209 @@ import (
+
+func nodeStatusError(err error, nodeNames ...string) *observerpb.GetFlowsResponse {
+	msg := err.Error()
+	if s, ok := status.FromError(err); ok && s.Code() == codes.Unknown {
```
Why only if `s.Code() == codes.Unknown`?

gandro replied:
It's mirroring the logic we have on the client-side: 

https://github.com/cilium/hubble/blob/9bca65beb242f467d869f044f54b58cf26e14c88/cmd/observe/observe.go#L90-L94

Unknown gRPC codes are errors thrown by the Hubble Observer for application-level errors such as `invalid IP address in filter`. Stripping out the `grpc error: code = Unknown desc =` makes this easier to understand for users, showing them that this is a problem with the request, not with grpc. 

All other codes (e.g. Unimplemented) tend to be protocol-level errors, therefore it makes sense to preserve the information that this is coming from the gRPC stack.

rolinh commented on pkg/hubble/relay/observer.go:
```diff
@@ -20,17 +20,209 @@ import (
+				// either merge with pending or flush it
+				if pending := pendingResponse.GetNodeStatus(); pending != nil {
+					if current.GetMessage() == pending.GetMessage() {
+						pending.NodeNames = append(pending.NodeNames, current.NodeNames...)
```
As I understand it, there are no node names de-dup logic, right?

gandro replied:
Yes. However, as it is implemented right now, a node can throw an error only once during a request. There cannot be any duplicate nodes to begin with.

rolinh replied:
Ah, right because there is no retry logic yet.

rolinh reviewed (approved):
I've tested this quite extensively and I'm pretty happy with this new API change and implementation. Well done!
